pkg/plugins/mellanox/mellanox_plugin.go

```diff
@@ -192,46 +222,33 @@ func getMlnxNicFwData(pciAddress string) (current, next *mlnxNic, err error) {
 		return
 	}
 	mstCurrentData, mstNextData := parseMstconfigOutput(out, attrs)
-	current, err = mlnxNicFromMap(mstCurrentData, firstPort)
+	current, err = mlnxNicFromMap(mstCurrentData)
 	if err != nil {
 		glog.Errorf("mellanox-plugin getMlnxNicFwData(): %v", err)
 		return
 	}
-	next, err = mlnxNicFromMap(mstNextData, firstPort)
+	next, err = mlnxNicFromMap(mstNextData)
 	if err != nil {
 		glog.Errorf("mellanox-plugin getMlnxNicFwData(): %v", err)
 	}
 	return
 }
 
-func mlnxNicFromMap(mstData map[string]string, firstPort bool) (*mlnxNic, error) {
-	glog.Infof("mellanox-plugin mlnxNicFromMap() %v, %v", mstData, firstPort)
-	fwData := &mlnxNic{firstPort: firstPort}
-	var linkType string
+func mlnxNicFromMap(mstData map[string]string) (*mlnxNic, error) {
+	glog.Infof("mellanox-plugin mlnxNicFromMap() %v", mstData)
+	fwData := &mlnxNic{}
 	if strings.Contains(mstData[EnableSriov], "True") {
 		fwData.enableSriov = true
 	}
 	i, err := strconv.Atoi(mstData[TotalVfs])
 	if err != nil {
 		return nil, err
 	}
-	fwData.totalVfs = i
-	if fwData.firstPort {
-		linkType = mstData[LinkTypeP1]
-	} else {
-		linkType = mstData[LinkTypeP2]
-	}
 
-	if strings.Contains(linkType, EthLinkType) {
-		fwData.linkType = EthLinkType
-	} else if strings.Contains(linkType, InfinibandLinkType) {
-		fwData.linkType = InfinibandLinkType
-	} else if len(linkType) > 0 {
-		glog.Warningf("mellanox-plugin getMlnxNicFwData(): link type %s is not one of [ETH, IB]", linkType)
-		fwData.linkType = UknownLinkType
-	} else {
-		glog.Warning("mellanox-plugin getMlnxNicFwData(): LINK_TYPE_P* attribute was not found")
-		fwData.linkType = PreconfiguredLinkType
+	fwData.totalVfs = i
+	fwData.linkTypeP1 = getLinkType(mstData[LinkTypeP1])
```

Comment:
'LINK_TYPE_P*' attributes do not alway existing in all the Mellanox NIC. I suppose the code will break if the there is no such attributes. Have you tested that?

Reply:
It is the same code, just add a function `getLinkType` for it, and regarding LINK_TYPE_P* not exits it is already checked in the last case in the function `getLinkType`